internal/cache/file/cache_handle.go

```diff
@@ -0,0 +1,114 @@
+// Copyright 2023 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package file
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"time"
+
+	"github.com/googlecloudplatform/gcsfuse/internal/cache/data"
+	"github.com/googlecloudplatform/gcsfuse/internal/cache/file/downloader"
+	"github.com/googlecloudplatform/gcsfuse/internal/cache/lru"
+	"github.com/googlecloudplatform/gcsfuse/internal/storage/gcs"
+)
+
+const (
+	InvalidFileHandle      = "fileHandle is nil"
+	InvalidFileDownloadJob = "download job is nil"
+	InvalidFileInfoCache   = "fileInfo cache is nil"
+	InvalidFileInfo        = "fileInfo is nil"
+)
+
+type CacheHandle struct {
+	// fileHandle to a local file which contains locally downloaded data.
+	fileHandle *os.File
+
+	//	fileDownloadJob is a reference to async download Job.
+	fileDownloadJob *downloader.Job
+
+	// Contains the latest information about the downloaded bits for a particular fileInfoKey.
+	fileInfoCache *lru.Cache
+}
+
+// validateCacheHandle will validate the  cache-handle and return appropriate error.
+func (fch *CacheHandle) validateCacheHandle() error {
+
+	if fch.fileHandle == nil {
+		return errors.New(InvalidFileHandle)
+	}
+
+	if fch.fileDownloadJob == nil {
+		return errors.New(InvalidFileDownloadJob)
+	}
+
+	if fch.fileInfoCache == nil {
+		return errors.New(InvalidFileInfoCache)
+	}
+
+	return nil
+}
+
+// Read attempts to read the data from the cached location. This expects a
+// fileInfoCache entry for the current read request, and will wait to download
+// the requested chunk if it is not already present.
+func (fch *CacheHandle) Read(object *gcs.MinObject, bucket gcs.Bucket, offset uint64, dst []byte) (n int, err error) {
```

Comment:
Please also check if we can handle the below cases now/when download job is ready:
(We are going to check file cache in all cases irrespective of type of read).
1. When all sequential reads turns into random for a file, we need to cancel the ongoing job.
2. When random read and --download-file-random-read is not set.
3. Not required immediately but it's possible that random read doesn't give good performance, in that case, we may want to trigger the download but not wait for it

Reply:
`When all sequential reads turns into random for a file, we need to cancel the ongoing job.` - As discussed, we will provide this method in CacheHandler, which will cancel the job for the specific fileInfoCache entry.

Reply:
Handled,  but disabled now, will enable when the download method is ready.